### PR TITLE
http2 client: enable http2 connection health check

### LINKED ISSUE #87615: (1.17) Kubelet won't reconnect to Apiserver after NIC failure (use of closed network connection)
We've just upgrade our production cluster to 1.17.2.

Since the update on saturday, we've had this strange outage: Kubelet, after a NIC bond fail (that recovers not long after), will have all of its connections broken and won't retry to restablish them unless manually restarted.

Here is the timeline of last time it occured:

01:31:16: Kernel recognizes a fail on the bond interface. It goes for a while. Eventually it recovers.
```
Jan 28 01:31:16 baremetal044 kernel: bond-mngmt: link status definitely down for interface eno1, disabling it
...
Jan 28 01:31:37 baremetal044  systemd-networkd[1702]: bond-mngmt: Lost carrier
Jan 28 01:31:37 baremetal044  systemd-networkd[1702]: bond-mngmt: Gained carrier
Jan 28 01:31:37 baremetal044  systemd-networkd[1702]: bond-mngmt: Configured
```

As expected, all watches are closed. Message is the same for them all:

```
...
Jan 28 01:31:44 baremetal044 kubelet-wrapper[2039]: W0128 04:31:44.352736    2039 reflector.go:326] object-"namespace"/"default-token-fjzcz": watch of *v1.Secret ended with: very short watch: object-"namespace"/"default-token-fjzcz": Unexpected watch close - watch lasted less than a second and no items received
...
```

So these messages begin:

```
`Jan 28 01:31:44 baremetal44 kubelet-wrapper[2039]: E0128 04:31:44.361582 2039 desired_state_of_world_populator.go:320] Error processing volume "disco-arquivo" for pod "pod-bb8854ddb-xkwm9_namespace(8151bfdc-ec91-48d4-9170-383f5070933f)": error processing PVC namespace/disco-arquivo: failed to fetch PVC from API server: Get https://apiserver:443/api/v1/namespaces/namespace/persistentvolumeclaims/disco-arquivo: write tcp baremetal44.ip:42518->10.79.32.131:443: use of closed network connection`
``` 

Which I'm guessing shouldn't be a problem for a while. But it never recovers. Our event came to happen at 01:31 AM, and had to manually restart Kubelet around 9h to get stuff normalized.

```
# journalctl --since '2020-01-28 01:31'   | fgrep 'use of closed' | cut -f3 -d' ' | cut -f1 -d1 -d':' | sort | uniq -dc
   9757 01
  20663 02
  20622 03
  20651 04
  20664 05
  20666 06
  20664 07
  20661 08
  16655 09
      3 10
```

Apiservers were up and running, all other nodes were up and running, everything else pretty uneventful. This one was the only one affected (today) by this problem.

Is there any way to mitigate this kind of event?

Would this be a bug?

### PR DESCRIPTION
**Pull request by Einsfier**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
There are several confirmed client connection issues with Go's http2 implementation.
And [it's also proved](https://github.com/kubernetes/kubernetes/issues/87615#issuecomment-671814091) that enabling http2 connection health check helps kubernetes client recover from such issues.

This PR updates golang.org/x/net to required version and enables http2 health check by default.

**Which issue(s) this PR fixes**:
Fixes #87615 #91963 #92164

**Special notes for your reviewer**:
Do we need a release note for this ?

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```
/area dependency
/sig api-machinery

/cc @liggitt @JensErat

### CODE REVIEW

**Comment by aojea** [comment]
:thinking: it looks a duplicate of https://github.com/kubernetes/kubernetes/pull/94844

**Comment by dims** [comment]
@povsister - @aojea is right. please check the other one out and ensure that fixes your problem as well.

/close

**Comment by povsister** [comment]
> 🤔 it looks a duplicate of #94844

Not really, original solution proposed in #94844 uses a modified http2 package which breaks backward compatibility and is not merged into golang/net master.
And he didn't push another fix till I proposed this fix.

To be honest(no offense), code and commit history in #94844 looks nasty...

**Comment by dims** [comment]
@povsister please comment on your concerns in the other PR. that is already approved and is likely to merge first.

**Comment by povsister** [comment]
Thx, I am

**Comment by aojea** [comment]
I think it doesn' t hurt if we test it in the meantime :sweat_smile: 
/ok-to-test

**Comment by povsister** [comment]
@aojea Thanks, seem one test job got OOMKilled. I will re-trigger it later.

**Comment by duyanghao** [comment]
> > 🤔 it looks a duplicate of #94844
> 
> Not really, original solution proposed in #94844 uses a modified http2 package which breaks backward compatibility and is not merged into golang/net master.
> And he didn't push another fix till I proposed this fix.
> 
> To be honest(no offense), code and commit history in #94844 looks nasty...

@povsister I don't know why you're so hysterical about which PR should be merged, and what I want to clarify here is that obviously I found this problem before you, and I have even not noticed the existence of golang/net@08b3837 until @fisherxu mentioned it in my PR, that why I push a new commit using http2.ConfigureTransports today and open this [PR](https://github.com/golang/net/pull/84) trying to fix this problem.

**Comment by povsister** [comment]
@duyanghao Calm down. I do not mind whose PR get merged. We are talking things technically, don't make it personal. 
If you insist. I'd like to tell you that [we have already talked about this issue months ago](https://github.com/kubernetes/kubernetes/issues/87615#issuecomment-647915537).

I prefer forward-looking solution instead of solving questions at hand. That's why I have opposition to your PR.

I have every concern with cause and conclusion posted in you PR, that's technical.
You have disagreement but no reason. That's personal.

If you feel disrespect on "nasty", I apologize. I have neat freak on coding : )
And you should learn how to use Git with clear&clean commit history.

**Comment by aojea** [comment]
This is not a competition and please respect the code of conduct
https://kubernetes.io/community/code-of-conduct/#contributor-code-of-conduct

There are certain comments in this thread that are clearly uncalled-for.

**Comment by povsister** [comment]
Looks like the update to golang.org/x/sys package breaks azure disk driver?

 It says `event for azuredisk-volume-tester-pbml9-84db65759d-9p2vq: {kubelet 9131k8s000} FailedMount: Unable to attach or mount volumes: unmounted volumes=[test-volume-1], unattached volumes=[test-volume-1 default-token-klq5d]: timed out waiting for the condition`

Code problem should not cause operation timeout. I think it maybe overload of test infra. I will retry next morning.

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 1]
```diff
@@ -154,6 +155,43 @@ func allowsHTTP2(t *http.Transport) bool {
...
+// to use golang.org/x/net HTTP/2 instead of bundled HTTP/2.
+// It gives us an opportunity to configure the underlying HTTP/2 Transport.
+// It will return error if the HTTP/1 Transport has already been HTTP/2-enabled.
+func configureHttp2Transport(t1 *http.Transport) error {
+	h2t, err := http2.ConfigureTransports(t1)
+	if err != nil {
+		return err
+	}
+
+	// There are several confirmed flaws with Go's http2 implementation.
```
Comment (JensErat) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:168]: I guess the flaws are in the specific networks. I agree golang's http2 keepalive implementation might have better defaults, but calling it a flaw seems rather harsh to me. "fatal situations" on the other hand hits the point. :+1: for references regarding why this is important!

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 2]
```diff
@@ -154,6 +155,43 @@ func allowsHTTP2(t *http.Transport) bool {
...
+		// h2Transport configured here does not dial any new connection,
+		// we can set the interval of healthCheck to half of h1Transport's IdleConnTimeout.
+		// By default, ReadIdleTimeout is not configured.
+		h2t.ReadIdleTimeout = idleConnTimeout
+	} else {
+		// The default IdleConnTimeout for h1Transport is 90s
+		h2t.ReadIdleTimeout = defaultTransport.IdleConnTimeout / 2
+	}
+	// The default PingTimeout is 15s.
+	// Our configuration should not below default value.
```
Comment (JensErat) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:187]: Nits: I think the comment is missing some word here.

```suggestion
	// Our configuration should not be below the default value.
```

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 3]
```diff
@@ -154,6 +155,43 @@ func allowsHTTP2(t *http.Transport) bool {
...
+	// There are several confirmed flaws with Go's http2 implementation.
+	// So we need to explicitly enable the "http2: connection health check" introduced in golang/net/pull/55
+	// to recover from such "fatal" situations.
+	// Upstream golang issues:
+	//   # golang/go/issues/39750
+	//   # golang/go/issues/40201
+	//   # golang/go/issues/41721
+	// Associated Kubernetes issues:
+	//   # kubernetes/kubernetes/issues/87615
+	if idleConnTimeout := t1.IdleConnTimeout / 2; idleConnTimeout > 0 {
```
Comment (JensErat) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:177]: Is there any argument on going for half of `idleConnTimeout` here? On first sight, this seems equally compared with @duyanghao's 10s or our own 30s interval. Not opting for any specific value here, but the calculation seems to interfere more thoughts on a proper value than obvious from the code.

**Review by JensErat — COMMENTED** [review]
Mostly /lgtm from my side, although prow will probably not care about it anyway.

Comparing this to #94844, I like that this more closely to sticking to the defaults. At Daimler, our workaround was patching x/net to keep `PingTimeout` default at 15 seconds but enforce a `ReadIdleTimeout` of 30 seconds. I guess this is sufficient for Kubernetes work loads, and at the same time should not really trigger "additional" keep-alive efforts in production too often. If I get it right, this PR sets the default to 90/2=45 seconds, and `t1.IdleConnTimeout` is never set in the entire Kubernetes organization to anything but the default values. I'm still not able to provide an argument to use half of `t1.IdleConnTimeout`, but our setting of 30 seconds was a random guess either. This should be triggering in rare error cases anyway.

#94844's timeouts seem a little harsh. At least for clusters spread around the world or having higher latency connections (Kubernetes going to the edge?), 2s seems very narrow.

Regarding the changes apart from the constants set, my preference towards this PR is probably mostly based on the extended discussion of why we need those "weird setup routine". Even if this is starting to get close to being over-commented, I'd even consider adding the note from [kubernetes/client-go#374](https://github.com/kubernetes/client-go/issues/374#issuecomment-717394606), which seems the best approach of describing the root cause (apart from the lost connection on layers 2 or 3) so far to me.

Maybe golang developers change their mind in future regarding ["reasonable defaults"](https://github.com/prometheus/prometheus/issues/7588#issuecomment-659376169) (I totally agree with @brian-brazil from Prometheus here), this explanation will help at deciding on whether to tear down this setup routine again.

Jens Erat <jens.erat@daimler.com>, Daimler TSS GmbH, [Imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go (hunk 3 above)
Comment (povsister) [reply staging/src/k8s.io/apimachinery/pkg/util/net/http.go:177]: Yes, I explained [why it should be calculated here](https://github.com/kubernetes/kubernetes/pull/94844#issuecomment-717226431). I mis-typed the local varible `readIdleTimeout` to `idleConnTimeout` causing some confusion here.

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go (hunk 2 above)
Comment (povsister) [reply staging/src/k8s.io/apimachinery/pkg/util/net/http.go:187]: Thx. Will fix later.

**Comment by aojea** [comment]
> At Daimler, our workaround was patching x/net to keep `PingTimeout` default at 15 seconds but enforce a `ReadIdleTimeout` of 30 seconds. I guess this is sufficient for Kubernetes work loads, and at the same time should not really trigger "additional" keep-alive efforts in production too often. 

> If I get it right, this PR sets the default to 90/2=45 seconds, and `t1.IdleConnTimeout` is never set in the entire Kubernetes organization to anything but the default values.


maybe it is me, but I prefer hardcoded timeouts than formulas, is not easy to me to understand which values I'm using this way.
Personally, as I user, I'd like to use the defaults timeouts provided or the ones I set directly, is my mistake if I set them wrong, but I found weird that I set `pingTimeout` to 3 seconds for whatever reason and something configures it to 15 seconds without I'm noticing.

**Comment by povsister** [comment]
> If I get it right, this PR sets the default to 90/2=45 seconds, and t1.IdleConnTimeout is never set in the entire Kubernetes organization to anything but the default values.

By default Kubernetes always set `t1.IdleConnTimeout` to default 90s. I added the if-compare section just for possible exceptions in the future, it ensures functionality of healthCheck.

> Personally, as I user, I'd like to use the defaults timeouts provided or the ones I set directly, is my mistake if I set them wrong.

For now, http2 Transport configuration is purely internal. We do not expose such configuration to users, that's probably why we don't need release note.
So I think we'd better make it reasonable even with some confusing formula.(still can be explained by sufficient comments in code, It should not be a big problem.)

**Comment by duyanghao** [comment]
> @duyanghao Calm down. I do not mind whose PR get merged. We are talking things technically, don't make it personal.
> If you insist. I'd like to tell you that [we have already talked about this issue months ago](https://github.com/kubernetes/kubernetes/issues/87615#issuecomment-647915537).
> 
> I prefer forward-looking solution instead of solving questions at hand. That's why I have opposition to your PR.
> 
> I have every concern with cause and conclusion posted in you PR, that's technical.
> You have disagreement but no reason. That's personal.
> 
> If you feel disrespect on "nasty", I apologize. I have neat freak on coding : )
> And you should learn how to use Git with clear&clean commit history.

@povsister You are so unbelievably arrogant that I don't want to waste my time arguing with you. And I am able to do git sqaush.

**Comment by liggitt** [comment]
@povsister @duyanghao thank you both for working to resolve the network connection issue. However, the discussion in this PR and in #94844 is distracting from the main goal of fixing the issue, and is violating the [Kubernetes code of conduct](https://kubernetes.io/community/code-of-conduct/) with insulting/derogatory comments. I've locked the discussion here and in #94844 for that reason, and will ask @caesarxuchao to take a look at the approaches in the two PRs and settle on a way to make use of the PingTimeout capability he added.

**Comment by lavalamp** [comment]
This isn't how we have technical disagreements. No amount of technical correctness can make up for treating other people poorly; please don't repeat that behavior. I will close both of these PRs. I'm available via slack or email if more explanation is needed.

Instead, I've asked @caesarxuchao to send a fix for the issue in question.

**Pull request closed without merging**